### PR TITLE
fortios_api module

### PR DESCRIPTION
**Pull request by willwagner602**
##### SUMMARY
Added fortios_api module to manage Fortinet device configuration REST API, including a utility class that will be used for further modules accessing specific Fortinet device endpoints.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
fortios_api

##### ANSIBLE VERSION
```
ansible 2.5.0 (fortios_api b236f8027f) last updated 2017/09/22 11:27:39 (GMT -400)
  config file = None
  configured module search path = [u'/home/wbw/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/wbw/Desktop/Ansible/lib/ansible
  executable location = /home/wbw/Desktop/Ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]

```

##### ADDITIONAL INFORMATION
The base module for a set of modules that will cover a wide range of Fortinet REST API functionality.

### CODE REVIEW

**Comment by mattclay** [comment]
CI failure in unit tests (in addition to the already noted sanity test failures):

https://app.shippable.com/github/ansible/ansible/runs/38427/2/tests

**Comment by eoprede** [comment]
Hello,
Any feedback? It looks like all the tests are passing.
And just to be sure - this is one module of over 20 that we are looking to submit, looking for suggestions on how to handle that best.
Thanks!

**Comment by gundalow** [comment]
@bjolivot I believe you created the original fortios modules, would be good to get your input on this

File: lib/ansible/modules/network/fortios/fortios_api.py
Comment (gundalow) [inline lib/ansible/modules/network/fortios/fortios_api.py]: Indentation looks wrong here and in `conn_params`

Comment (eoprede) [reply lib/ansible/modules/network/fortios/fortios_api.py]: Thanks, that's fixed now.

**Comment by bjolivot** [comment]
Hello
Great work @willwagner602
using Fortios API is probably a more reliable way to deal with fortios devices than using pyFG library.

This PR is looks good to me.
shipit

**Comment by eoprede** [comment]
@gundalow The unit tests of the most recent build work fine on my laptop, but they fail in Shippable due to lack of requests_mock module. Is it possible to instruct Shippable to install this module, or do tests need to be changed?
Thanks!

**Comment by mattclay** [comment]
CI failure in unit tests:

```
ImportError while importing test module '/root/src/github.com/ansible/ansible/test/units/modules/network/fortios/test_fortios_api.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
test/units/modules/network/fortios/test_fortios_api.py:6: in <module>
    import requests_mock
E   ImportError: No module named requests_mock
```

File: lib/ansible/module_utils/fortios.py
Comment (gundalow) [inline lib/ansible/module_utils/fortios.py]: Why has this been removed?

File: lib/ansible/modules/network/fortios/fortios_api.py
Comment (gundalow) [inline lib/ansible/modules/network/fortios/fortios_api.py:108]: In `RETURN` blocks we now support nested returns (eg returning complex data structures, see http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#return-block

File: lib/ansible/module_utils/fortios.py
Comment (gundalow) [inline lib/ansible/module_utils/fortios.py]: Ansible coding guidelines allow up to 160 chars (which is enforced by CI)
Please revert these line-length changes it as
1) It's not needed
2) It makes the diff "noisy"

File: lib/ansible/utils/module_docs_fragments/fortios_api_doc.py
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/fortios_api_doc.py]: Does Fortigate support different levels of user privileges? If so please state what privileges this user needs to have. 
e.g "Specifies username for Fortigate connection with admin privileges"

File: lib/ansible/utils/module_docs_fragments/fortios_api_doc.py
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/fortios_api_doc.py]: Please remove this and add a note in the `description:` saying the default port will be based on the connection type, e.g. 80 or 443 for http or https respectively.

File: lib/ansible/utils/module_docs_fragments/fortios_api_doc.py
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/fortios_api_doc.py]: `, must be located in conn_params dict` can be removed here, and in the following lines.

the `suboptions:` dictionary will create a nested table with the options, e.g. http://docs.ansible.com/ansible/latest/vyos_config_module.html#options

File: lib/ansible/utils/module_docs_fragments/fortios_api_doc.py
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/fortios_api_doc.py]: `C(api-path-CurrentConfig.json)` for formatting

File: lib/ansible/utils/module_docs_fragments/fortios_api_doc.py
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/fortios_api_doc.py]: `to use  I(print_current_config: true)`

**Review by gundalow — CHANGES_REQUESTED** [review]

**Comment by gundalow** [comment]
I've added a note about the python module onto https://github.com/ansible/community/issues/248, will discuss there and get back to you.

File: lib/ansible/modules/network/fortios/fortios_api.py
Comment (eoprede) [reply lib/ansible/modules/network/fortios/fortios_api.py:108]: Thanks for the info. I am not entirely sure if there's any benefit to using complex data structures in documentation. Also, the returns are different per different API endpoint, so samples will vary greatly (and I think in some other modules we don't even include them, as the response may be a few pages long)

**Comment by gundalow** [comment]
@bjolivot Could you please review this again, thanks in advance

**Comment by sivel** [comment]
I'll provide a general comments here in regards to tests.  This info was discussed in the testing meeting today.

All of the unit tests should be converted to use pytest and not rely on unittest as we are moving away from unittest.

I'd also like to see us move to using `responses` instead of `requests_mock` which seems to be more used throughout the community. In addition to this, I would question the uses of either `responses` or `requests_mock`.  Since our modules are not permitted to use `requests` directly, there should typically be no need to mock `requests` at all.  Instead you should be mocking functions that call out of the function you are testing.  We don't need to test the python libraries that we are using.

For example:

```
def some_function():
    response = some_function_that_calls_outside_of_this_function()
    # something with response here
```

You should mock `some_function_that_calls_outside_of_this_function` rather than mocking any use of requests within `some_function_that_calls_outside_of_this_function`.

In addition to switching to using pytest, this will facilitate a need to move some things directly into test functions and use mocking decorators where possible instead of overloading so much into the `setUp` function.

I'll try to make some other specific comments inline with the code as well in the near future.

File: lib/ansible/module_utils/fortios.py
Comment (sivel) [inline lib/ansible/module_utils/fortios.py:53]: Module are not permitted to uses the requests module, due to unnecessarily adding dependencies to modules.  You should utilize the functionality in `ansible.module_utils.urls` instead of requests.

**Review by sivel — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/fortios.py
Comment (eoprede) [reply lib/ansible/module_utils/fortios.py:53]: One of the reasons for using requests proxy was that it easily supports socks5 proxy, which is a requirement in our environment. It seems that the urls module doesn't support it now ("Currently ansible only supports HTTP proxies"), so unfortunately dependency on requests (or any other external library to handle HTTP connections) is required.
If there's a way to use socks proxies with ansible built-in modules, could you please share how that can be done?

Comment (sivel) [reply lib/ansible/module_utils/fortios.py:53]: That functionality should be added to `urls.py`

Comment (eoprede) [reply lib/ansible/module_utils/fortios.py:53]: I am afraid that my limited coding capabilities will not allow me to add socks proxy support to the urls.py module in a reasonable timeframe. Is lack of dependency on requests a hard requirement for the module to be accepted by Ansible?

Comment (sivel) [reply lib/ansible/module_utils/fortios.py:53]: Is the dependency on a socks proxy part of your specific environment, or something that every person using this module will need?

I have more reviewing to do, but I would be fundamentally against the module having specific settings for socks if it isn’t a requirement of fortios.

Just as with using other proxies it should be a fundamental part of how Ansible can be used and should not be a determining factor of how a module is implemented due to one persons environmental restrictions.

Comment (eoprede) [reply lib/ansible/module_utils/fortios.py:53]: The dependency on socks proxy is requirement of my specific environment (to be specific, it's some of the devices in my environment, others don't need any proxy to connect to them), but it is important to note that my employer is paying me salary for managing this specific environment.
My employer is understanding the importance of open source and is ok with me spending extra time to make this module part of ansible (especially since working on documentation and tests is making the code better), but I am afraid that adding the code to urls.py and then changing fortios.py to achieve exactly what the module is already doing successfully would be a stretch for my employer. And unfortunately due to my relative weakness of coding and personal reasons, I will not be able to devote sufficient personal time to solve this issue in the near future.
Would it be possible to have module rely on requests library for the time being and then use the power of open source community to make necessary adjustments to urls.py and this module to make it not rely on the  requests library?

File: test/units/modules/network/fortios/test_fortios_api.py
Comment (sivel) [inline test/units/modules/network/fortios/test_fortios_api.py:1]: See this comment for information about changes needed to this file:

https://github.com/ansible/ansible/pull/30767#issuecomment-339752582

File: test/units/modules/network/fortios/test_fortios_api.py
Comment (sivel) [inline test/units/modules/network/fortios/test_fortios_api.py:22]: You should just be able to use `mock_open` and patching the location of where open is used, instead of determining the correct location of the builtin to mock.

File: test/units/modules/network/fortios/test_fortios_api.py
Comment (sivel) [inline test/units/modules/network/fortios/test_fortios_api.py:38]: These `open` calls are not safe, and should likely be done inside of `with open(...)` to ensure the file is closed properly.

File: test/units/modules/network/fortios/test_fortios_api.py
Comment (sivel) [inline test/units/modules/network/fortios/test_fortios_api.py:83]: These are really awkward line breaks. Ansible uses a max width of 160 chars, this is not necessary. This applies pretty much throughout the code base.

File: test/units/modules/network/fortios/test_fortios_api.py
Comment (sivel) [inline test/units/modules/network/fortios/test_fortios_api.py:100]: Should just be using printf style formatting in full here instead of a mix of both `+` concatenation and printf formatting

File: test/units/modules/network/fortios/test_fortios_api.py
Comment (sivel) [inline test/units/modules/network/fortios/test_fortios_api.py:114]: Another example of awkward line breaks.

File: test/units/modules/network/fortios/test_fortios_api.py
Comment (sivel) [inline test/units/modules/network/fortios/test_fortios_api.py:779]: Another example of awkward line breaks.

File: lib/ansible/module_utils/fortios.py
Comment (sivel) [inline lib/ansible/module_utils/fortios.py:615]: I'm not fully sure what this is yet, but I would recommend renaming this function, as it is likely to cause confusion for devs/user due to ansible already calling something an "argument_spec"

File: lib/ansible/module_utils/fortios.py
Comment (sivel) [inline lib/ansible/module_utils/fortios.py:595]: What is the need here for writing and reading from a JSON file?  Seems this file would be created during execution, but in a temporary directory, and deleted immediately.  Not sure what benefit there is to doing this, as opposed to just storing in a variable for use.

File: lib/ansible/module_utils/fortios.py
Comment (sivel) [inline lib/ansible/module_utils/fortios.py:531]: I highly recommend using printf style formatting or using `.format()` instead of `+` concatenation. There is a performance hit for python to try and understand if it can do the operation by type inspecting the parts, figuring out how to do it, and what to do in the case of a failure.

File: lib/ansible/module_utils/fortios.py
Comment (sivel) [inline lib/ansible/module_utils/fortios.py:307]: I don't believe this parameter should exist.  I believe we should rely on the ability for libraries to use environment variables for `http_proxy` and `https_proxy`.  Both `ansible.module_utils.urls` and `requests` can both utilize these environment vars.

Setting these values via the `environment` keyword on a task is accepted.  The module should not have a deviating method for applying proxies.

File: lib/ansible/module_utils/fortios.py
Comment (sivel) [inline lib/ansible/module_utils/fortios.py:514]: You seem to have gone through the process of making this class a context manager, but don't use it as a context manager.

File: lib/ansible/module_utils/fortios.py
Comment (sivel) [inline lib/ansible/module_utils/fortios.py:919]: I'm slightly confused by why functions like `_lists_match`, `_dictionaries_match` and `_unknown_types_match` are needed.

Python can already natively compare these things. Is it due to order?

File: lib/ansible/module_utils/fortios.py
Comment (sivel) [inline lib/ansible/module_utils/fortios.py:45]: Why aren't we able to continue forward using `pyFG` as opposed to writing all the interactions ourselves?

**Review by sivel — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/fortios.py
Comment (eoprede) [reply lib/ansible/module_utils/fortios.py:45]: Because pyFG is using SSH scraping and I wanted to use API, as it is a better way of managing devices.

File: lib/ansible/module_utils/fortios.py
Comment (eoprede) [reply lib/ansible/module_utils/fortios.py:595]: This file will be kept locally and will not be deleted. It's to reduce amount of API calls during the execution.

Comment (sivel) [reply lib/ansible/module_utils/fortios.py:595]: How is that?

```
self._arg_spec_filename = "FortiosAPIArgSpecs.json"
```

That defines the path to that filename as relative, which will be relative to the executing script.  modules are executed in a temp dir that is then deleted, so this file should be deleted.

If the file is kept around, I would be concerned with that.  This is force creating a cache file without telling the user you are doing it, or giving them a way to opt out.  The default should be no cache, and a user must opt-in, should this feature really need to exist.

File: lib/ansible/module_utils/fortios.py
Comment (eoprede) [reply lib/ansible/module_utils/fortios.py:514]: The only thing I want to do is to logout when the execution is complete or failed. I believe both __del__ and __exit__ are being called and to avoid having duplicate logout calls I just ended up doing it in __del__ (line 507)

File: lib/ansible/module_utils/fortios.py
Comment (eoprede) [reply lib/ansible/module_utils/fortios.py:615]: This function is making an API call to get schema for the API endpoint and makes an argument spec for the given endpoint out of it. For example:
    "cmdb/firewall/addrgrp": {
        "arg_spec": {
            "allow-routing": {
                "options": [
                    "enable", 
                    "disable"
                ], 
                "type": "str"
            }, 
            "color": {
                "type": "int"
            }, 
            "comment": {
                "type": "str"
            }, 
            "member": {
                "options": {
                    "name": {
                        "type": "str"
                    }
                }, 
                "type": "list"
            }, 
            "name": {
                "type": "str"
            }, 
            "tags": {
                "options": {
                    "name": {
                        "type": "str"
                    }
                }, 
                "type": "list"
            }, 
            "uuid": {
                "type": "str"
            }, 
            "visibility": {
                "options": [
                    "enable", 
                    "disable"
                ], 
                "type": "str"
            }
        }, 
        "default_object": {
            "allow-routing": "disable", 
            "color": 0, 
            "comment": "", 
            "css-class": "ftnt-address-group ftnt-color-0", 
            "member": [], 
            "name": "", 
            "q_origin_key": "", 
            "tags": [], 
            "uuid": "00000000-0000-0000-0000-000000000000", 
            "visibility": "enable"
        }
    },

File: lib/ansible/module_utils/fortios.py
Comment (eoprede) [reply lib/ansible/module_utils/fortios.py:307]: Just want to verify - will it be possible to apply these environment variables per host in the inventory file? The idea is that you may have multiple firewalls, some of which need proxy (or even different proxies) to be managed and others don't.

Comment (sivel) [reply lib/ansible/module_utils/fortios.py:307]: This should be modeled in the same way that all other modules within ansible currently work.  Variables can be set per host and applied to tasks.

File: lib/ansible/module_utils/fortios.py
Comment (eoprede) [reply lib/ansible/module_utils/fortios.py:595]: Currently this file is created in the folder that you are running the playbook from. I can make the opt-in feature.

Comment (sivel) [reply lib/ansible/module_utils/fortios.py:595]: I think that may be a result of running the module in the way you are.  But it shouldn't be.  Imagine you aren't running the module "locally" but running it on a remote machine.

File: lib/ansible/module_utils/fortios.py
Comment (sivel) [reply lib/ansible/module_utils/fortios.py:615]: I think that makes me even more concerned.  You are saying that the argument spec that ansible will use becomes dynamic based on responses of APIs?

I'm pretty sure that's not going to be allowed.

Comment (sivel) [reply lib/ansible/module_utils/fortios.py:615]: We have historically not allowed generic API modules.  We generally require a module to be a specific set of features that perform an action.  This would be in opposition to that stand point.

Comment (eoprede) [reply lib/ansible/module_utils/fortios.py:615]: That is correct, argument spec is dynamic. If it is not dynamic, then any change to the API would render this module useless. Also, it would require me to write modules for every single API endpoint. There are some specific modules - https://github.com/willwagner602/ansible/tree/will-devel/lib/ansible/modules/network/fortios but there are a lot more of endpoints. So it made sense to let this core module build dynamic argument spec based on the data received from the API.

Comment (eoprede) [reply lib/ansible/module_utils/fortios.py:615]: And just to clarify - currently the argument spec per API endpoint is being cached locally. So if API changes, the file has to be deleted and then it can be recreated again.
Here's an example of this file after running a few playbooks: https://pastebin.com/CcPuMQCu

Comment (sivel) [reply lib/ansible/module_utils/fortios.py:615]: Due to this, at the moment I am denying acceptance of this PR.  I'll ask to get additional feedback from some other core developers, but we have explicitly denied modules that attempted to address arbitrary API functionality like this in the past.

The recommended way would be to write individual modules for individual parts of functionality.

Comment (eoprede) [reply lib/ansible/module_utils/fortios.py:615]: OK, I feel like it makes sense for the modules that are using screen scraping via SSH, but I see a lot of value of treating API calls as generic as possible. I'll hold off on making any changes to the module until the core members provide feedback. I'll be happy to attend the meeting to explain my point of view to the core developers.

**Comment by sivel** [comment]
In alignment with other module submissions of this type, and based on discussions with members of the core team, and what is available as documentation here within the review, we are going to decline accepting this pull request.

The primary driver is that it is an arbitrary API module, that has no definable argument spec.

Each module should typically interact with one set of functionality.  Should you decide to break this module up into individual components that deal with one set of functionalities with the API that have an explicitly defined argument spec, we would be happy to re-review.

**Pull request closed without merging**
